File: sdk/go/client/resilience.go
package client

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/protocol"
)

const ErrorCodeIdempotentInFlight protocol.ErrorCode = "IDEMPOTENT_IN_FLIGHT"

// RetryPolicy controls the thin retry decorator. It retries handler-level
// TaskResult failures only; panics are left for the lifecycle guard.
type RetryPolicy struct {
	MaxAttempts       int
	InitialDelay      time.Duration
	BackoffMultiplier float64
	ShouldRetry       func(TaskResult) bool
}

func DefaultRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxAttempts:       3,
		InitialDelay:      100 * time.Millisecond,
		BackoffMultiplier: 2,
		ShouldRetry: func(result TaskResult) bool {
			switch result.ErrorCode {
			case protocol.ErrorCodeExecutionFailed, protocol.ErrorCodeTimeout, protocol.ErrorCodeResourceExhausted:
				return true
			default:
				return false
			}
		},
	}
}

// WithRetry wraps a handler with a result-based retry policy.
func WithRetry(handler TaskHandler, policy RetryPolicy) TaskHandler {
	if handler == nil {
		panic("nil TaskHandler")
	}
	if policy.MaxAttempts < 1 {
		policy.MaxAttempts = 1
	}
	if policy.BackoffMultiplier <= 0 {
		policy.BackoffMultiplier = 1
	}
	if policy.ShouldRetry == nil {
		policy.ShouldRetry = DefaultRetryPolicy().ShouldRetry
	}
	return retryHandler{inner: handler, policy: policy}
}

type retryHandler struct {
	inner  TaskHandler
	policy RetryPolicy
}

func (h retryHandler) Execute(ctx *TaskContext) TaskResult {
	delay := h.policy.InitialDelay
	var result TaskResult
	for attempt := 1; attempt <= h.policy.MaxAttempts; attempt++ {
		result = h.inner.Execute(ctx)
		if result.IsSuccess() || attempt == h.policy.MaxAttempts || !h.policy.ShouldRetry(result) {
			return result
		}
		if delay > 0 {
			time.Sleep(delay)
			delay = time.Duration(float64(delay) * h.policy.BackoffMultiplier)
		}
	}
	return result
}

// IdempotencyEntity is the cached success snapshot returned by the idempotency
// decorator when the same key is replayed.
type IdempotencyEntity struct {
	Result TaskResult
}

// SdkIdempotencyStore is intentionally storage-agnostic. Production tenants can
// back it with DB/Redis; the SDK ships only noop and in-memory test stores.
type SdkIdempotencyStore interface {
	TryAcquire(ctx context.Context, key string, ttl time.Duration) (bool, error)
	Find(ctx context.Context, key string) (IdempotencyEntity, bool, error)
	Record(ctx context.Context, key string, entity IdempotencyEntity, ttl time.Duration) error
	Release(ctx context.Context, key string) error
}

type IdempotencyKeyFunc func(*TaskContext) (string, error)

func TaskIDIdempotencyKey(ctx *TaskContext) (string, error) {
	if ctx == nil || ctx.TaskID == "" {
		return "", fmt.Errorf("taskId is required")
	}
	return ctx.TaskID, nil
}

// WithIdempotency wraps a handler with acquire/cache/release semantics. Only
// successful results are recorded; failures release the in-flight marker.
func WithIdempotency(handler TaskHandler, store SdkIdempotencyStore, ttl time.Duration, keyFunc IdempotencyKeyFunc) TaskHandler {
	if handler == nil {
		panic("nil TaskHandler")
	}
	if store == nil {
		return handler
	}
	if keyFunc == nil {
		keyFunc = TaskIDIdempotencyKey
	}
	return idempotentHandler{inner: handler, store: store, ttl: ttl, keyFunc: keyFunc}
}

type idempotentHandler struct {
	inner   TaskHandler
	store   SdkIdempotencyStore
	ttl     time.Duration
	keyFunc IdempotencyKeyFunc
}

func (h idempotentHandler) Execute(taskCtx *TaskContext) (result TaskResult) {
	key, err := h.keyFunc(taskCtx)
	if err != nil || key == "" {
		if err == nil {
			err = fmt.Errorf("idempotency key is empty")
		}
		return Fail(protocol.ErrorCodeConfigInvalid, err.Error())
	}
	ctx := context.Background()
	if taskCtx != nil && taskCtx.Cancellation != nil {
		ctx = taskCtx.Cancellation.Context()
	}
	acquired, err := h.store.TryAcquire(ctx, key, h.ttl)
	if err != nil {
		return Fail(protocol.ErrorCodeExecutionFailed, "idempotency acquire failed: "+err.Error())
	}
	if !acquired {
		entity, found, err := h.store.Find(ctx, key)
		if err != nil {
			return Fail(protocol.ErrorCodeExecutionFailed, "idempotency lookup failed: "+err.Error())
		}
		if found {
			return entity.Result
		}
		return Fail(ErrorCodeIdempotentInFlight, "idempotent execution is already in flight")
	}

	defer func() {
		if r := recover(); r != nil {
			_ = h.store.Release(ctx, key)
			panic(r)
		}
	}()

	result = h.inner.Execute(taskCtx)
	if result.IsSuccess() {
		if err := h.store.Record(ctx, key, IdempotencyEntity{Result: result}, h.ttl); err != nil {
			return Fail(protocol.ErrorCodeExecutionFailed, "idempotency record failed: "+err.Error())
		}
		return result
	}
	_ = h.store.Release(ctx, key)
	return result
}

type NoopIdempotencyStore struct{}

func (NoopIdempotencyStore) TryAcquire(context.Context, string, time.Duration) (bool, error) {
	return true, nil
}

func (NoopIdempotencyStore) Find(context.Context, string) (IdempotencyEntity, bool, error) {
	return IdempotencyEntity{}, false, nil
}

func (NoopIdempotencyStore) Record(context.Context, string, IdempotencyEntity, time.Duration) error {
	return nil
}

func (NoopIdempotencyStore) Release(context.Context, string) error {
	return nil
}

type InMemoryIdempotencyStore struct {
	mu      sync.Mutex
	entries map[string]idempotencyEntry
	now     func() time.Time
}

type idempotencyEntry struct {
	entity  *IdempotencyEntity
	expires time.Time
}

func NewInMemoryIdempotencyStore() *InMemoryIdempotencyStore {
	return &InMemoryIdempotencyStore{
		entries: map[string]idempotencyEntry{},
		now:     time.Now,
	}
}

func (s *InMemoryIdempotencyStore) TryAcquire(_ context.Context, key string, ttl time.Duration) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.deleteIfExpiredLocked(key)
	if _, exists := s.entries[key]; exists {
		return false, nil
	}
	s.entries[key] = idempotencyEntry{expires: expiresAt(s.now(), ttl)}
	return true, nil
}

func (s *InMemoryIdempotencyStore) Find(_ context.Context, key string) (IdempotencyEntity, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.deleteIfExpiredLocked(key)
	entry, exists := s.entries[key]
	if !exists || entry.entity == nil {
		return IdempotencyEntity{}, false, nil
	}
	return *entry.entity, true, nil
}

func (s *InMemoryIdempotencyStore) Record(_ context.Context, key string, entity IdempotencyEntity, ttl time.Duration) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries[key] = idempotencyEntry{entity: &entity, expires: expiresAt(s.now(), ttl)}
	return nil
}

func (s *InMemoryIdempotencyStore) Release(_ context.Context, key string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.entries, key)
	return nil
}

func (s *InMemoryIdempotencyStore) deleteIfExpiredLocked(key string) {
	entry, exists := s.entries[key]
	if !exists || entry.expires.IsZero() || s.now().Before(entry.expires) {
		return
	}
	delete(s.entries, key)
}

func expiresAt(now time.Time, ttl time.Duration) time.Time {
	if ttl <= 0 {
		return time.Time{}
	}
	return now.Add(ttl)
}

File: sdk/rust/src/client/resilience.rs
use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use crate::client::handler::{TaskContext, TaskHandler, TaskResult};
use crate::protocol::error_code;

pub const IDEMPOTENT_IN_FLIGHT: &str = "IDEMPOTENT_IN_FLIGHT";

pub struct RetryPolicy {
    pub max_attempts: usize,
    pub initial_delay: Duration,
    pub backoff_multiplier: f64,
    pub should_retry: fn(&TaskResult) -> bool,
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self {
            max_attempts: 3,
            initial_delay: Duration::from_millis(100),
            backoff_multiplier: 2.0,
            should_retry: default_should_retry,
        }
    }
}

pub fn default_should_retry(result: &TaskResult) -> bool {
    matches!(
        result.error_code.as_str(),
        error_code::EXECUTION_FAILED | error_code::TIMEOUT | error_code::RESOURCE_EXHAUSTED
    )
}

pub fn with_retry<H: TaskHandler>(inner: H, policy: RetryPolicy) -> RetryHandler<H> {
    RetryHandler { inner, policy }
}

pub struct RetryHandler<H: TaskHandler> {
    inner: H,
    policy: RetryPolicy,
}

impl<H: TaskHandler> TaskHandler for RetryHandler<H> {
    fn task_type(&self) -> &str {
        self.inner.task_type()
    }

    fn execute(&self, ctx: &TaskContext) -> TaskResult {
        let max_attempts = self.policy.max_attempts.max(1);
        let multiplier = if self.policy.backoff_multiplier > 0.0 {
            self.policy.backoff_multiplier
        } else {
            1.0
        };
        let mut delay = self.policy.initial_delay;
        let mut result = TaskResult::failure(error_code::EXECUTION_FAILED, "not executed");

        for attempt in 1..=max_attempts {
            result = self.inner.execute(ctx);
            if result.is_success()
                || attempt == max_attempts
                || !(self.policy.should_retry)(&result)
            {
                return result;
            }
            if !delay.is_zero() {
                std::thread::sleep(delay);
                delay = Duration::from_millis(((delay.as_millis() as f64) * multiplier) as u64);
            }
        }
        result
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IdempotencyEntity {
    pub result: TaskResult,
}

pub trait SdkIdempotencyStore: Send + Sync {
    fn try_acquire(&self, key: &str, ttl: Duration) -> Result<bool, String>;
    fn find(&self, key: &str) -> Result<Option<IdempotencyEntity>, String>;
    fn record(&self, key: &str, entity: IdempotencyEntity, ttl: Duration) -> Result<(), String>;
    fn release(&self, key: &str) -> Result<(), String>;
}

pub fn task_id_idempotency_key(ctx: &TaskContext) -> Result<String, String> {
    if ctx.task_id.is_empty() {
        return Err("taskId is required".to_string());
    }
    Ok(ctx.task_id.clone())
}

pub fn with_idempotency<H, F>(
    inner: H,
    store: Arc<dyn SdkIdempotencyStore>,
    ttl: Duration,
    key_resolver: F,
) -> IdempotencyHandler<H, F>
where
    H: TaskHandler,
    F: Fn(&TaskContext) -> Result<String, String> + Send + Sync,
{
    IdempotencyHandler {
        inner,
        store,
        ttl,
        key_resolver,
    }
}

pub struct IdempotencyHandler<H, F>
where
    H: TaskHandler,
    F: Fn(&TaskContext) -> Result<String, String> + Send + Sync,
{
    inner: H,
    store: Arc<dyn SdkIdempotencyStore>,
    ttl: Duration,
    key_resolver: F,
}

impl<H, F> TaskHandler for IdempotencyHandler<H, F>
where
    H: TaskHandler,
    F: Fn(&TaskContext) -> Result<String, String> + Send + Sync,
{
    fn task_type(&self) -> &str {
        self.inner.task_type()
    }

    fn execute(&self, ctx: &TaskContext) -> TaskResult {
        let key = match (self.key_resolver)(ctx) {
            Ok(key) if !key.is_empty() => key,
            Ok(_) => {
                return TaskResult::failure(error_code::CONFIG_INVALID, "idempotency key is empty")
            }
            Err(e) => return TaskResult::failure(error_code::CONFIG_INVALID, &e),
        };

        let acquired = match self.store.try_acquire(&key, self.ttl) {
            Ok(v) => v,
            Err(e) => {
                return TaskResult::failure(
                    error_code::EXECUTION_FAILED,
                    &format!("idempotency acquire failed: {e}"),
                )
            }
        };

        if !acquired {
            return match self.store.find(&key) {
                Ok(Some(entity)) => entity.result,
                Ok(None) => TaskResult::failure(
                    IDEMPOTENT_IN_FLIGHT,
                    "idempotent execution is already in flight",
                ),
                Err(e) => TaskResult::failure(
                    error_code::EXECUTION_FAILED,
                    &format!("idempotency lookup failed: {e}"),
                ),
            };
        }

        let result = self.inner.execute(ctx);
        if result.is_success() {
            if let Err(e) = self.store.record(
                &key,
                IdempotencyEntity {
                    result: result.clone(),
                },
                self.ttl,
            ) {
                return TaskResult::failure(
                    error_code::EXECUTION_FAILED,
                    &format!("idempotency record failed: {e}"),
                );
            }
            result
        } else {
            let _ = self.store.release(&key);
            result
        }
    }
}

#[derive(Default)]
pub struct NoopIdempotencyStore;

impl SdkIdempotencyStore for NoopIdempotencyStore {
    fn try_acquire(&self, _key: &str, _ttl: Duration) -> Result<bool, String> {
        Ok(true)
    }

    fn find(&self, _key: &str) -> Result<Option<IdempotencyEntity>, String> {
        Ok(None)
    }

    fn record(&self, _key: &str, _entity: IdempotencyEntity, _ttl: Duration) -> Result<(), String> {
        Ok(())
    }

    fn release(&self, _key: &str) -> Result<(), String> {
        Ok(())
    }
}

#[derive(Default)]
pub struct InMemoryIdempotencyStore {
    entries: Mutex<BTreeMap<String, IdempotencyEntry>>,
}

#[derive(Clone)]
struct IdempotencyEntry {
    entity: Option<IdempotencyEntity>,
    expires_at: Option<Instant>,
}

impl InMemoryIdempotencyStore {
    pub fn new() -> Self {
        Self::default()
    }

    fn delete_if_expired_locked(entries: &mut BTreeMap<String, IdempotencyEntry>, key: &str) {
        let expired = entries
            .get(key)
            .and_then(|entry| entry.expires_at)
            .map(|expires_at| Instant::now() >= expires_at)
            .unwrap_or(false);
        if expired {
            entries.remove(key);
        }
    }
}

impl SdkIdempotencyStore for InMemoryIdempotencyStore {
    fn try_acquire(&self, key: &str, ttl: Duration) -> Result<bool, String> {
        let mut entries = self.entries.lock().map_err(|e| e.to_string())?;
        Self::delete_if_expired_locked(&mut entries, key);
        if entries.contains_key(key) {
            return Ok(false);
        }
        entries.insert(
            key.to_string(),
            IdempotencyEntry {
                entity: None,
                expires_at: expires_at(ttl),
            },
        );
        Ok(true)
    }

    fn find(&self, key: &str) -> Result<Option<IdempotencyEntity>, String> {
        let mut entries = self.entries.lock().map_err(|e| e.to_string())?;
        Self::delete_if_expired_locked(&mut entries, key);
        Ok(entries.get(key).and_then(|entry| entry.entity.clone()))
    }

    fn record(&self, key: &str, entity: IdempotencyEntity, ttl: Duration) -> Result<(), String> {
        let mut entries = self.entries.lock().map_err(|e| e.to_string())?;
        entries.insert(
            key.to_string(),
            IdempotencyEntry {
                entity: Some(entity),
                expires_at: expires_at(ttl),
            },
        );
        Ok(())
    }

    fn release(&self, key: &str) -> Result<(), String> {
        let mut entries = self.entries.lock().map_err(|e| e.to_string())?;
        entries.remove(key);
        Ok(())
    }
}

fn expires_at(ttl: Duration) -> Option<Instant> {
    if ttl.is_zero() {
        None
    } else {
        Some(Instant::now() + ttl)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicUsize, Ordering};

    struct CountingHandler {
        calls: Arc<AtomicUsize>,
    }

    impl TaskHandler for CountingHandler {
        fn task_type(&self) -> &str {
            "demo"
        }

        fn execute(&self, _ctx: &TaskContext) -> TaskResult {
            let call = self.calls.fetch_add(1, Ordering::SeqCst) + 1;
            if call < 3 {
                TaskResult::failure(error_code::EXECUTION_FAILED, "temporary")
            } else {
                TaskResult::success("ok").with_output("attempts", &call.to_string())
            }
        }
    }

    #[test]
    fn retry_retries_retryable_result() {
        let calls = Arc::new(AtomicUsize::new(0));
        let handler = with_retry(
            CountingHandler {
                calls: Arc::clone(&calls),
            },
            RetryPolicy {
                max_attempts: 3,
                initial_delay: Duration::ZERO,
                ..RetryPolicy::default()
            },
        );

        let result = handler.execute(&TaskContext::new("t1", "tenant", "demo"));

        assert!(result.is_success());
        assert_eq!(calls.load(Ordering::SeqCst), 3);
        assert_eq!(
            result.outputs.get("attempts").map(String::as_str),
            Some("3")
        );
    }

    #[test]
    fn retry_stops_on_non_retryable_result() {
        struct BadConfig;
        impl TaskHandler for BadConfig {
            fn task_type(&self) -> &str {
                "demo"
            }
            fn execute(&self, _ctx: &TaskContext) -> TaskResult {
                TaskResult::failure(error_code::CONFIG_INVALID, "bad config")
            }
        }

        let result = with_retry(
            BadConfig,
            RetryPolicy {
                max_attempts: 3,
                initial_delay: Duration::ZERO,
                ..RetryPolicy::default()
            },
        )
        .execute(&TaskContext::new("t1", "tenant", "demo"));

        assert_eq!(result.error_code, error_code::CONFIG_INVALID);
    }

    #[test]
    fn idempotency_caches_success() {
        struct Once {
            calls: Arc<AtomicUsize>,
        }
        impl TaskHandler for Once {
            fn task_type(&self) -> &str {
                "demo"
            }
            fn execute(&self, _ctx: &TaskContext) -> TaskResult {
                let call = self.calls.fetch_add(1, Ordering::SeqCst) + 1;
                TaskResult::success("done").with_output("calls", &call.to_string())
            }
        }

        let calls = Arc::new(AtomicUsize::new(0));
        let handler = with_idempotency(
            Once {
                calls: Arc::clone(&calls),
            },
            Arc::new(InMemoryIdempotencyStore::new()),
            Duration::from_secs(60),
            task_id_idempotency_key,
        );
        let ctx = TaskContext::new("same", "tenant", "demo");

        let first = handler.execute(&ctx);
        let second = handler.execute(&ctx);

        assert!(first.is_success());
        assert!(second.is_success());
        assert_eq!(calls.load(Ordering::SeqCst), 1);
        assert_eq!(second.outputs.get("calls").map(String::as_str), Some("1"));
    }

    #[test]
    fn idempotency_releases_after_failure() {
        struct FailsOnce {
            calls: Arc<AtomicUsize>,
        }
        impl TaskHandler for FailsOnce {
            fn task_type(&self) -> &str {
                "demo"
            }
            fn execute(&self, _ctx: &TaskContext) -> TaskResult {
                let call = self.calls.fetch_add(1, Ordering::SeqCst) + 1;
                if call == 1 {
                    TaskResult::failure(error_code::EXECUTION_FAILED, "temporary")
                } else {
                    TaskResult::success("ok")
                }
            }
        }

        let calls = Arc::new(AtomicUsize::new(0));
        let handler = with_idempotency(
            FailsOnce {
                calls: Arc::clone(&calls),
            },
            Arc::new(InMemoryIdempotencyStore::new()),
            Duration::from_secs(60),
            task_id_idempotency_key,
        );
        let ctx = TaskContext::new("retry", "tenant", "demo");

        assert!(!handler.execute(&ctx).is_success());
        assert!(handler.execute(&ctx).is_success());
        assert_eq!(calls.load(Ordering::SeqCst), 2);
    }

    #[test]
    fn idempotency_reports_in_flight_without_snapshot() {
        struct Never;
        impl TaskHandler for Never {
            fn task_type(&self) -> &str {
                "demo"
            }
            fn execute(&self, _ctx: &TaskContext) -> TaskResult {
                panic!("handler should not run")
            }
        }

        let store = Arc::new(InMemoryIdempotencyStore::new());
        assert!(store.try_acquire("busy", Duration::from_secs(60)).unwrap());
        let handler = with_idempotency(Never, store, Duration::from_secs(60), |_ctx| {
            Ok("busy".to_string())
        });

        let result = handler.execute(&TaskContext::new("t1", "tenant", "demo"));

        assert_eq!(result.error_code, IDEMPOTENT_IN_FLIGHT);
    }
}
